File: LeetCode/3. Longest Substring Without Repeating Characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        left, right = 0, 0
        max_len = 0
        seen = set()
        while right < len(s):
            if s[right] in seen:
                seen.remove(s[left])
                left += 1
            else:
                seen.add(s[right])
                right += 1
                max_len = max(max_len, right - left)
        return max_len

File: LeetCode/3. Longest Substring Without Repeating Characters/Solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> map = new HashMap<>();
        int left = 0, right = 0;
        int maxlen = 0;
        
        while (right < s.length()) {
            if (map.containsKey(s.charAt(right))) {
                left = Math.max(left, map.get(s.charAt(right)) + 1);
            }
            map.put(s.charAt(right), right);
            right++;
            maxlen = Math.max(maxlen, right - left);
        }
        return maxlen;
    }
}
